src/js/websocket.js

```diff
@@ -0,0 +1,199 @@
+      reason: msg_str.substr(4, msg_str.length),
+    };
+  } else {
+    msg = {};
+  }
+
+  this.client.emit('close', msg);
+  for (var i = 0; i < this.pings.length; i++) {
+    clearInterval(this.pings[i].timer);
+  }
+  this.client._socket.end();
+};
+
+WebSocketHandle.prototype.sendFrame = function(msg, cb) {
+  if (this.connected) {
+    if (typeof cb == 'function') {
+      this.client._socket.write(msg, cb);
+    } else {
+      this.client._socket.write(msg);
+    }
+  } else {
+    this.onError('Underlying socket connection is closed');
+  }
+};
+
+WebSocketHandle.prototype.pong = function(msg) {
+  this.client._socket.write(native.ping(false, msg, true));
+};
+
+WebSocketHandle.prototype.onpingresp = function(msg) {
+  for (var i = 0; i < this.pings.length; i++) {
+    if (this.pings[i].id == msg) {
+      clearInterval(this.pings[i].timer);
+      this.pings[i].callback(msg);
+      this.pings.splice(i, 1);
+      return;
+    }
+  }
+};
+
+function sendHandshake(jsref, host, path) {
+  return native.prepareHandshake(jsref, host, path);
+}
+
+Websocket.prototype.connect = function(url, port, path, callback) {
+  var host = url.toString() || '127.0.0.1';
+  path = path || '/';
+
+  var emit_type = 'connect';
+
+  if (host.substr(0, 3) == 'wss') {
+    this._secure = true;
+    if (!tls) {
+      this._handle.onError('TLS module was not found!');
+    }
+    port = port || 443;
+    host = host.substr(6);
+    this._socket = tls.connect(port, host);
+    emit_type = 'secureConnect';
+  } else if (host.substr(0, 2) == 'ws') {
+    port = port || 80;
+    this._socket = new net.Socket();
+    host = host.substr(5);
+  } else {
+    port = port || 80;
+    this._socket = new net.Socket();
+  }
+
+  if (typeof callback == 'function') {
+    this.on('open', callback);
+  }
+
+  var self = this;
+
+  this._socket.on(emit_type, function() {
+    self._handle.connected = true;
+    self._socket.write(sendHandshake(self._handle, host, path));
+  });
+
+  this._socket.on('end', function() {
+    self._handle.connected = false;
+  });
+  if (emit_type == 'connect') {
+    this._socket.connect(port, host);
+  }
+
+  this._socket.on('data', function(data) {
+    if (self._firstMessage) {
+      var remaining_data = native.parseHandshakeData(data);
+      self._handle.onhandshakedone(remaining_data);
+    } else {
+      self._handle.ondata(data);
+    }
+  });
+};
+
+Websocket.prototype.close = function(message, code, cb) {
+  this._handle.sendFrame(native.close(message, code), cb);
+};
+
+Websocket.prototype.ping = function(message, mask, cb) {
+  var self = this;
+  var obj = {
+    id: message,
+    callback: cb,
+    timer: setTimeout(function() {
+      self.close('Ping timeout limit exceeded', 1002);
+    }, 30000),
+  };
+  this._handle.pings.push(obj);
+  this._handle.sendFrame(native.ping(true, message, mask));
+};
+
+Websocket.prototype.send = function(message, opts, cb) {
+  if (opts) {
+    var mask = opts.mask ? true : false;
```

Comment:
Every JS value has a logical value, and this cannot throw an error. Just pass the value.